community/contributing/back-end-code.md

```diff
@@ -0,0 +1,75 @@
+Interfaces, for example require almost no changes.
+Otherwise, changes to existing Groovy code are fine, but any new classes should not be written in Groovy.
+
+Types from the Groovy runtime libraries should not be exposed in the API of any class.
+Since Groovy closures can be automatically type-coerced to Java [https://dzone.com/articles/java-8-functional-interfaces-sam](SAM types), please use an appropriate SAM type for parameters or return types that may be implemented with Groovy closures.
+
+### Java
+
+In any module (i.e. a sub-project of an microservice) that already uses Groovy we prefer new code be written in Java.
+
+Spinnaker uses Java 8 and we encourage use of lambdas, the streams API, etc. where appropriate.
+
+### Kotlin
+
+In any module that does _not_ currently use Groovy the language of choice is [https://kotlinlang.org/](Kotlin).
+However, it is absolutely fine to write Java code if that's what you're more comfortable with.
+
+There are cross-compilation issues with mixing Groovy and Kotlin in the same source tree.
+For that reason, and in the interests of retaining some measure of sanity, we would rather not mix both languages in a single module.
+
+## Code formatting
+
+We follow [https://google.github.io/styleguide/javaguide.html](Google's Java Style Guide).
+
+For Kotlin / Groovy languages please use:
+
+* 2 space indents.
+* No more than 1 consecutive line of whitespace.
+* Line breaks rather than overly long lines.
+* Camel case conventions as per Java.
+
+## Package structure
+
+Spinnaker microservices automatically component-scan for `@Configuration` classes in the `com.netflix.spinnaker.<service>.config` package.
+Other classes should be placed in `com.netflix.spinnaker.<service>.<feature>` where `<service>` is the microservice name, for example `orca` or `clouddriver` and `<feature>` is something descriptive of the feature being implemented.
+
+Please do not separate classes into different packages by _what_ they are.
+Packages should represent the group of classes that implement a particular piece of functionality not all components of a particular type.
+
+## Naming things
+
+Please use descriptive but concise names for variables, classes, properties and so on.
+Longer names are fine when they add clarity.
+
+### A note on units
+
+It's preferable to use types that properly represent things like durations, timestamps, etc. (`java.time.Duration` and `java.time.Instant` would be ideal in those specific cases).
+If that's not practical please include a suffix on the property / variable name that describes the unit.
+A property declared as `public long getTimeout()` is ambiguous and can easily lead to errors when developers using your code assume what the units are.
+
+For example, these names are much less likely to result in errors:
+```
+public long getTimeoutMillis();
+public long getTimeoutSeconds();
+```
```

Comment:
I know this is meant to be a start of the doc, and things we should add shortly are test coverage expectations, applicable documentation, etc.

Reply:
This point was actually my motivation for starting the doc!

Yes, I definitely agree we need to extend it and testing / documentation would be the first things to cover.